### PR TITLE
Enhancements of our Form

### PR DESCRIPTION
**Pull request by romaninsh**
It's time we focus on cleaning up Form a bit more, specifically with some Fields. This PR aims to solve number of issues:

 - #281 $field->set() should work
 - #290 improve demo
 - #155 checkbox saving

We will also add more UI tests (behat) and will try to switch ZombieJS with PhantomJS (#264).

### CODE REVIEW

**Comment by DarkSide666** [comment]
Some thoughts (ticked if fixed):
 - [ ] I don't think we need FormField/Line. It's the same as Input and I don't see any need for this class. Currently Password field extends it, but it could simply extend Input class too. Only advantage of having Input/Line as two classes I see is if we move part of decoration functionality from Input to Line and some other field classes like Money and TextArea (which have similar decoration capabilities) extends Line not Input. So Input will become more abstract class and Line will be nicely decorated Input.
 - [ ] Probably Radio don't work too because its implementation is quite similar to Checkbox
 - [ ] Methods jsInput(), getValue(), getInput(), set() and part of renderView() should be moved from Input to Generic
 - [ ] Maybe we need setValue() method too which will correctly typecast field value?

Implemented part of all that, but I don't like that messy code. Especially treating $this->content as field value and then there is template "Content" etc. That's a mess. Need to sort this out and make much cleaner and nicer codebase :)

File: src/FormField/CheckBox.php [hunk 1]
```diff
@@ -13,13 +13,26 @@ class CheckBox extends Generic
...
     public $defaultTemplate = 'formfield/checkbox.html';
 
-    public $label;
+    /**
+     * Label appears to the right of the checkbox. If label is not set specifically
+     * then the $caption property will be displayed as a label instead.
+     */
+    public $label = null;
+
+    public function __construct($label = null, $class = null)
```
Comment (romaninsh) [inline src/FormField/CheckBox.php]: with constructors it's normal not to call parent. The idea of a constructor is to simply set some default values. It should not be complex. We do not support array argument on purpose.

File: src/FormField/CheckBox.php [hunk 2]
```diff
@@ -31,10 +44,31 @@ public function init()
...
+        }
+
+        /*
+         * We don't want this displayed, because it can only affect "checked" status anyway
+         */
+        $this->content = null;
+
         $this->js(true)->checkbox();
 
+        $this->content = null; // no content again
```
Comment (romaninsh) [inline src/FormField/CheckBox.php]: Due to the way how template defines $Content we can still add a more sophisticated label (such as link or custom message) inside a label.

File: src/FormField/Input.php [hunk 3]
```diff
@@ -177,9 +180,4 @@ public function addAction($defaults = [])
 
         return $this->action;
     }
-
-    public function set($value = null, $junk = null)
```
Comment (romaninsh) [inline src/FormField/Input.php]: this has gone to Generic.

**Comment by romaninsh** [comment]
- Input field is essential. You will appreciate it more as we add more non-input fields in the future (e.g. capcha or multi-input fields such as "day/month/year)
 - disagree on moving stuff to Generic. Removed some of them, because not all fields will support it or it may be too confusing.
 - setValue() - not needed. set() will call model->set which will typecast, once https://github.com/atk4/data/issues/262 is fixed.
 - content is value. I've tried moving it back and fourth but this is probably the best. I have added comments for Caption and Label to avoid confusion.

**Pull request merged**